### ISSUE TITLE
Support of DbContext pooling with token (Azure SQL). Missing by design?

### ISSUE THREAD

**Issue by BenjaminAbt** [dotnet/efcore#24034]
We use Azure SQL and the Managed Identity of our App Service to connect with our EFCore 5 App.

```cs
                services.AddDbContextPool<MyMssqlDbContext>(o =>
                {
                    o.UseSqlServer(databaseConnectionstring);
                    o.AddInterceptors(new AzureActiveDirectoryAuthenticationDbConnectionInterceptor());

                    if (databaseLoggingEnabled) { o.UseLoggerFactory(DatabaseDebugLoggerFactory); }
                });
```

Due to the Managed Identity we enrich the SQL connection with an Access Token based on https://mderriey.com/2020/07/17/connect-to-azure-sql-with-aad-and-managed-identities/

```cs
                AccessToken accessToken = await GetAzureSqlAccessToken(cancellationToken).ConfigureAwaitFalse();
                sqlConnection.AccessToken = accessToken.Token;
```` 

The following error now appears in the logs again and again:

```
Microsoft.Data.SqlClient.SqlException (0x80131904): Login failed for user ''.
   at Microsoft.Data.ProviderBase.DbConnectionPool.CheckPoolBlockingPeriod(Exception e)
   at Microsoft.Data.ProviderBase.DbConnectionPool.CreateObject(DbConnection owningObject, DbConnectionOptions userOptions, DbConnectionInternal oldConnection)
   at Microsoft.Data.ProviderBase.DbConnectionPool.UserCreateRequest(DbConnection owningObject, DbConnectionOptions userOptions, DbConnectionInternal oldConnection)
   at Microsoft.Data.ProviderBase.DbConnectionPool.TryGetConnection(DbConnection owningObject, UInt32 waitForMultipleObjectsTimeout, Boolean allowCreate, Boolean onlyOneCheckConnection, DbConnectionOptions userOptions, DbConnectionInternal& connection)
   at Microsoft.Data.ProviderBase.DbConnectionPool.TryGetConnection(DbConnection owningObject, TaskCompletionSource`1 retry, DbConnectionOptions userOptions, DbConnectionInternal& connection)
   at Microsoft.Data.ProviderBase.DbConnectionFactory.TryGetConnection(DbConnection owningConnection, TaskCompletionSource`1 retry, DbConnectionOptions userOptions, DbConnectionInternal oldConnection, DbConnectionInternal& connection)
   at Microsoft.Data.ProviderBase.DbConnectionInternal.TryOpenConnectionInternal(DbConnection outerConnection, DbConnectionFactory connectionFactory, TaskCompletionSource`1 retry, DbConnectionOptions userOptions)
   at Microsoft.Data.SqlClient.SqlConnection.TryOpen(TaskCompletionSource`1 retry, SqlConnectionOverrides overrides)
   at Microsoft.Data.SqlClient.SqlConnection.Open(SqlConnectionOverrides overrides)
   at Microsoft.EntityFrameworkCore.Storage.RelationalConnection.OpenInternal(Boolean errorsExpected)
```

Because of the error message and because it looked at first that it occurs only sometimes, the troubleshooting was now very time-consuming. But we could then see that the error occurs periodically every 5 minutes (not exactly), we could narrow that down to the token.

Due to the pooling the connection is reused, but it looks like the pool does not check the expiration of a token.
What I can see is that it works without those errors if I disable pooling.

Unfortunately I can't find any hint in the [documentation](https://docs.microsoft.com/en-us/ef/core/performance/advanced-performance-topics?tabs=with-constant#dbcontext-pooling) if DbContext Pooling supports this scenario (tokens) at all.
There is this 

> Context pooling works by reusing the same context instance across requests. This means that it's effectively registered as a Singleton in terms of the instance itself so that it's able to persist.

which could maybe mean this scenario is not supported, because a connection with a token "cannot" be a singleton.
But this is not a "state", but simply an expiration date of the connection. 

So I had assumed that the pool supports this scenario, but I can't find any check in the source code, too.
Was this forgotten or is it by design?

EF Core version: 5.0.1
Database provider: Microsoft.EntityFrameworkCore.SqlServer with Azure SQL
Target framework: 5.0.100
Operating system: App Service Linux

**Comment by roji** [comment]
@BenjaminAbt for background, DbContext pooling is about pooling the DbContext instances, and should not affect the connection interceptor being invoked or anything related to the token. The interceptor gets invoked every time EF opens the SqlConnection, and that shouldn't be affected by DbContext pooling in any way.

> But we could then see that the error occurs periodically every 5 minutes (not exactly), we could narrow that down to the token.

I may be misunderstanding you, but if the error occurs *once* around every 5 minutes, I think that means that the interceptor is indeed being invoked and the token is refreshed - but that there's one failed attempt for some reason, does that sound right?

Can you also confirm whether disabling DbContext pooling makes the problem go away entirely? If so, can you share a minimal project or code sample that shows how you're setting everything up?

**Comment by BenjaminAbt** [comment]
Hi, thank you very much for your answer.

Yes, disabling the pool fixed the problem, but the problem then came back. The intervals of the error message were still very irregular, occurred less often.

We have since fixed the problem by no longer using Managed Identity. So there can no longer be a condition where a token expires in context.

---

Unfortunately, I haven't had time to explore the problem in more depth, but since yesterday I no longer suspect EF Core. The fact is this:

We have an ASP.NET Core application that can be used both authenticated and anonymous. For authentication we use ASP.NET Core Identity in its original version without any customization.

In the meantime I could limit the error to one place:
- The problem only occurs when a non-authenticated user accesses a page that invokes a full-text search against Azure SQL.
- The error never shows up when the user is authenticated, nor when the SQL query does not use full-text search.
- The error occurs only when used with App Service for Linux.

The last point was then crucial that it can't be my code and not EF Core.
While searching for the problem related to App Service for Linux, I came across a few hits, for example:
https://github.com/Azure/azure-sdk-for-js/issues/11595 This is for JavaScript but I also found quite a few .NET hits.

--- 
Sorry for the thread, I think the bug needs to be reported elsewhere and is not related to pooling.
